File: COLLEGE/ASSIGNMENT6/python/Q.10.py
class MultiStack:
    def __init__(self, num_stacks, stack_size):
        self.num_stacks = num_stacks
        self.stack_size = stack_size
        self.array = [None] * (num_stacks * stack_size)
        self.sizes = [0] * num_stacks  # To keep track of each stack's current size

    def is_empty(self, stack_num):
        return self.sizes[stack_num] == 0

    def is_full(self, stack_num):
        return self.sizes[stack_num] == self.stack_size

    def index_of_top(self, stack_num):
        # Calculate the index of the top of the given stack
        offset = stack_num * self.stack_size
        return offset + self.sizes[stack_num] - 1

    def push(self, stack_num, value):
        if self.is_full(stack_num):
            raise OverflowError(f"Stack {stack_num} is full!")
        self.sizes[stack_num] += 1
        self.array[self.index_of_top(stack_num)] = value

    def pop(self, stack_num):
        if self.is_empty(stack_num):
            raise IndexError(f"Stack {stack_num} is empty!")
        top_index = self.index_of_top(stack_num)
        value = self.array[top_index]
        self.array[top_index] = None  # Clear the value
        self.sizes[stack_num] -= 1
        return value

    def peek(self, stack_num):
        if self.is_empty(stack_num):
            raise IndexError(f"Stack {stack_num} is empty!")
        return self.array[self.index_of_top(stack_num)]

# Test the MultiStack implementation
if __name__ == "__main__":
    num_stacks = 3
    stack_size = 5
    multi_stack = MultiStack(num_stacks, stack_size)

    # Push values to different stacks
    multi_stack.push(0, 10)
    multi_stack.push(0, 20)
    multi_stack.push(1, 30)
    multi_stack.push(2, 40)
    multi_stack.push(2, 50)

    # Peek values from different stacks
    print(f"Top of stack 0: {multi_stack.peek(0)}")  # Should print 20
    print(f"Top of stack 1: {multi_stack.peek(1)}")  # Should print 30
    print(f"Top of stack 2: {multi_stack.peek(2)}")  # Should print 50

    # Pop values from different stacks
    print(f"Popped from stack 0: {multi_stack.pop(0)}")  # Should print 20
    print(f"Popped from stack 2: {multi_stack.pop(2)}")  # Should print 50

File: COLLEGE/ASSIGNMENT6/c/Q.10.c
#include <stdio.h>
#include <stdlib.h>

#define MAX 100  // Total size of the array
#define STACK_COUNT 3  // Number of stacks

// Structure to represent multiple stacks
typedef struct {
    int arr[MAX];           // Array to store stack elements
    int top[STACK_COUNT];   // Array to track top elements of each stack
    int size;               // Size of each stack
} MultiStack;

// Function to initialize the multi-stack
void initialize(MultiStack *ms, int size) {
    ms->size = size;
    for (int i = 0; i < STACK_COUNT; i++) {
        ms->top[i] = -1;  // Initialize tops to -1 (indicating empty stacks)
    }
}

// Function to push an element onto a specific stack
void push(MultiStack *ms, int stackNum, int value) {
    if (stackNum < 0 || stackNum >= STACK_COUNT) {
        printf("Invalid stack number!\n");
        return;
    }
    
    if (ms->top[stackNum] + 1 >= ms->size) {
        printf("Stack %d overflow!\n", stackNum);
        return;
    }

    // Calculate actual position in the array
    int pos = stackNum * ms->size + (++ms->top[stackNum]);
    ms->arr[pos] = value;
    printf("Pushed %d onto stack %d\n", value, stackNum);
}

// Function to pop an element from a specific stack
int pop(MultiStack *ms, int stackNum) {
    if (stackNum < 0 || stackNum >= STACK_COUNT) {
        printf("Invalid stack number!\n");
        return -1;  // Return -1 to indicate an error
    }

    if (ms->top[stackNum] == -1) {
        printf("Stack %d underflow!\n", stackNum);
        return -1;  // Return -1 to indicate an error
    }

    // Calculate actual position in the array
    int pos = stackNum * ms->size + (ms->top[stackNum]--);
    return ms->arr[pos];
}

// Function to display the contents of a specific stack
void displayStack(MultiStack *ms, int stackNum) {
    if (stackNum < 0 || stackNum >= STACK_COUNT) {
        printf("Invalid stack number!\n");
        return;
    }

    printf("Stack %d: ", stackNum);
    for (int i = 0; i <= ms->top[stackNum]; i++) {
        printf("%d ", ms->arr[stackNum * ms->size + i]);
    }
    printf("\n");
}

// Main function to test the multiple stacks implementation
int main() {
    MultiStack ms;
    initialize(&ms, MAX / STACK_COUNT);  // Initialize with equal size for each stack

    // Push elements onto different stacks
    push(&ms, 0, 10);
    push(&ms, 0, 20);
    push(&ms, 1, 30);
    push(&ms, 1, 40);
    push(&ms, 2, 50);

    // Display stacks
    displayStack(&ms, 0);
    displayStack(&ms, 1);
    displayStack(&ms, 2);

    // Pop elements from stacks
    printf("Popped from stack 0: %d\n", pop(&ms, 0));
    printf("Popped from stack 1: %d\n", pop(&ms, 1));

    // Display stacks after popping
    displayStack(&ms, 0);
    displayStack(&ms, 1);
    displayStack(&ms, 2);

    return 0;
}
